apps/shipments/permissions.py

```diff
@@ -1,17 +1,66 @@
+from django.conf import settings
 from django.core.exceptions import ObjectDoesNotExist
-from rest_framework import permissions
+from rest_framework import permissions, status
 
+from apps.authentication import get_jwt_from_request
 from apps.permissions import has_owner_access
 from .models import PermissionLink, Shipment
 
 
+PROFILES_URL = f'{settings.PROFILES_URL}/api/v1/wallet'
+
+
 class IsShipmentOwner(permissions.BasePermission):
     """
     Custom permission to only allow owners of an object to edit it (for use in related querysets)
     """
+
+    def has_object_permission(self, request, view, obj):
+        shipment_id = request.parser_context['kwargs'].get('shipment_pk', None)
+
+        shipment = Shipment.objects.get(id=shipment_id)
+
+        return (has_owner_access(request, shipment) or self.is_shipper(request, shipment) or
+                self.is_carrier(request, shipment) or self.is_moderator(request, shipment))
+
     def has_permission(self, request, view):
-        shipment = Shipment.objects.get(pk=view.kwargs['shipment_pk'])
-        return has_owner_access(request, shipment)
+        shipment_id = request.parser_context['kwargs'].get('shipment_pk', None)
+        if not shipment_id:
+            return True
+
+        shipment = Shipment.objects.get(id=shipment_id)
+
+        return has_owner_access(request, shipment) or self.is_shipper(request, shipment) or \
```

Comment:
Most of the time we won't have access to the shipment object here. You will be better off moving this logic to the `has_object_permission`.